enhancements/art/flexibly-enforcing-product-build-constraints-in-ci.md

```diff
@@ -0,0 +1,297 @@
+       `FROM registry.svc.ci.openshift.org/ocp/builder:rhel-8-golang-1.14-openshift-4.6 AS builder`
+       to
+       `FROM registry.svc.ci.openshift.org/ocp/builder:rhel-8-golang-1.15-openshift-4.6 AS builder`
+    4. In response to these PRs, CI will build images using the updated golang and run all test jobs for that repo.
+    5. Production builds during this time will detect that many upstream components' Dockerfiles are using the older, 
+       but valid, ART equivalent 1.14 golang image. Instead of using the production metadata's desired state 
+       (golang 1.15) to compile, the previous version, golang 1.14, will be used to build those images.
+    6. Downstream releases which contain builds from inconsistent upstream builder images will be produced as 
+       normal nightlies, but they will be annotated in such a way as to make the release controller 
+       highlight the discrepancy. 
+4. **Day 19**: Half the upstream alignment PRs have been tested by component owners and merged. 
+    ART is automatically building those components now with the desired golang 1.15 and the remaining upstream 
+    components with golang 1.14.
+5. **Day 20**: Component X owner is unable to migrate to golang 1.15. 
+    1. They justify an exception with the architecture team and subsequently communicate the approved requirement to ART. 
+    2. ART places the component image on a golang stream independent from the control-plane. 
+    3. The alignment PR is updated automatically to reflect the new desired state. If that desired state 
+       is identical to Day 0, the alignment PR is a no-op and can be closed.
+6. **Day 45**: Feature freeze is approaching. ART queries for open alignment PRs and finds 2 in the search. 
+   Urgent BZs are opened against both components to communicate the need for either alignment or production metadata changes.
+7. **Day 48**: Feature freeze. ART changes production metadata to reflect the beginning of a new stage of the software 
+   lifecycle. One urgent BZ, despite escalation, remains unanswered. It is changed to a blocker bug. To ensure consistent 
+   and predictable builds leading up to and after GA, the production metadata is now authoritative and the upstream 
+   discrepancy is ignored (the PR and BZ remain open).
+8. **Day N**: If, at any point after feature freeze, the ART build pipeline detects that an upstream repository is not using 
+   the builder image defined in ART metadata, the release will be marked as inconsistent (visible in the release controller 
+   and programmatically accessible by checking an annotation). If the ART pipeline attempts to promote an inconsistent release, 
+   it will abort unless it is manually overridden. The manual override would only be undertaken if it was determined, 
+   with architectural approval, that the inconsistency was introduced erroneously and that the promotion is a critical 
+   delivery (e.g. a CVE fix).
```

Comment:
note that it is completely normal practice to stay on an old (previous) golang version (e.g. a repo without rebase) until some point after feature freeze. Then the need can occur to bump that repo to newer dependencies (neither due to CVE nor critical change, just a normal bug) and we have to move to new Golang version.

Why are we so concerned about these changes that we need architectural approval? I could understand this in a z-stream, but not before GA.

Reply:
Agreed. My mind was definitely on GA when writing the arch approval line. Added clarification.